In [1]:
import time
from cp import cp, main

TEST 1: Simple Test

This test only has 2 ingredients which the user already owns, and 1 recipe using the ingredients, so the solver will just pick the 1 recipe.

In [2]:
data1 = {
    "all_ingredients": [
        {"i_id": 0, "name": "Chicken Breast", "unit_price": 100},
        {"i_id": 1, "name": "Rice", "unit_price": 50},
    ],
    "recipes": [
        {
            "name": "Chicken and Rice",
            "ingredients": [{"i_id": 0, "name": "Chicken Breast", "proportion": 1.0},
                            {"i_id": 1, "name": "Rice", "proportion": 1.0}],
            "nutrients": {"calories": 400, "protein": 40, "cholesterol": 80}
        }
    ]
}

budget1 = 10 
calorie_cap1 = 500
inventory1 = [
    {"i_id": 0, "amount": 1.0}, 
    {"i_id": 1, "amount": 1.0},  
]
disliked_ct1 = [0]  # no dislikes
allergies = []
start = time.time()
result = cp(data1, budget1, calorie_cap1, inventory1, disliked_ct1, allergies, chosen_meals=1)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


Solution status: OPTIMAL
Chosen recipe IDs: [0]
  -> Chicken and Rice
Extra purchased amounts:
Total cost used: 0.0 (Budget = 10)
Total Protein: 40
Total Calories: 400
Total Cholestrol: 80
Calories to Protein Ratio: 10.0
Calories to Cholesterol Ratio: 5.0

execution time: 0.0202 seconds


TEST 2: No inventory

In this test, the user must buy the needed ingredients, note that they buy 3 units of beef despite the recipes calling for 2.7

In [3]:

data2 = {
    "all_ingredients": [
        {"i_id": 0, "name": "Beef", "unit_price": 200},
        {"i_id": 1, "name": "Potatoes", "unit_price": 75},
    ],
    "recipes": [
        {
            "name": "Beef Stew",
            "ingredients": [{"i_id": 0, "name": "Beef", "proportion": 1.5},
                            {"i_id": 1, "name": "Potatoes", "proportion": 2.0}],
            "nutrients": {"calories": 900, "protein": 50, "cholesterol": 120}
        },
        {
            "name": "Beef Brew",
            "ingredients": [{"i_id": 0, "name": "Beef", "proportion": 1.2}],
            "nutrients": {"calories": 600, "protein": 40, "cholesterol": 100}
        }
    ]
}

budget2 = 20
calorie_cap2 = 2000
inventory2 = [
    {"i_id": 0, "amount": 0.0}, 
    {"i_id": 1, "amount": 0.0},
]
disliked_ct2 = [0, 0]
allergies = []
start = time.time()
cp(data2, budget2, calorie_cap2, inventory2, disliked_ct2, allergies, chosen_meals=2)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


Solution status: OPTIMAL
Chosen recipe IDs: [0, 1]
  -> Beef Stew
  -> Beef Brew
Extra purchased amounts:
Ingredient 0 (Beef): 3.0
Ingredient 1 (Potatoes): 2.0
Total cost used: 7.5 (Budget = 20)
Total Protein: 90
Total Calories: 1500
Total Cholestrol: 220
Calories to Protein Ratio: 16.666666666666668
Calories to Cholesterol Ratio: 6.818181818181818

execution time: 0.0154 seconds


TEST 3, 4: Choose 2/3 Recipes

In the following tests, the user wants 2 out of the 3 available recipes but in the second test, they dislike pasta, thus changing the output.
In the first test, chicken pasta and pasta salad are selected because they maximize protein for roughly the same level of other macros, but in the second test, pasta salad is replaced with spinach salad because the user dislikes pasta. However, because chicken pasta is so high is protein, the user's soft constraint is overriden.

In [4]:
data3 = {
    "all_ingredients": [
        {"i_id": 0, "name": "Chicken", "unit_price": 100},
        {"i_id": 1, "name": "Pasta", "unit_price": 75},
        {"i_id": 2, "name": "Spinach", "unit_price": 125},
    ],
    "recipes": [
        {
            "name": "Chicken Pasta",
            "ingredients": [{"i_id": 0, "name": "Chicken", "proportion": 1.0},
                            {"i_id": 1, "name": "Pasta", "proportion": 1.0}],
            "nutrients": {"calories": 500, "protein": 50, "cholesterol": 100}
        },
        {
            "name": "Spinach Salad",
            "ingredients": [{"i_id": 2, "name": "Spinach", "proportion": 1.0}],
            "nutrients": {"calories": 200, "protein": 15, "cholesterol": 0}
        },
        {
            "name": "Pasta Salad",
            "ingredients": [{"i_id": 1, "name": "Pasta", "proportion": 0.5},
                            {"i_id": 2, "name": "Spinach", "proportion": 0.5}],
            "nutrients": {"calories": 200, "protein": 20, "cholesterol": 10}
        }
    ]
}

budget3 = 200
calorie_cap3 = 500
inventory3 = [
    {"i_id": 0, "amount": 1.0},
    {"i_id": 1, "amount": 1.0},
    {"i_id": 2, "amount": 1.0},
]
disliked_ct3 = [0, 0, 0]
allergies = []
start = time.time()
cp(data3, budget3, calorie_cap3, inventory3, disliked_ct3, allergies, chosen_meals=2)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


Solution status: OPTIMAL
Chosen recipe IDs: [0, 2]
  -> Chicken Pasta
  -> Pasta Salad
Extra purchased amounts:
Ingredient 1 (Pasta): 1.0
Ingredient 2 (Spinach): 1.0
Total cost used: 2.0 (Budget = 200)
Total Protein: 70
Total Calories: 700
Total Cholestrol: 110
Calories to Protein Ratio: 10.0
Calories to Cholesterol Ratio: 6.363636363636363

execution time: 0.0146 seconds


In [5]:
#here, assume we dislike pasta, meaning that recipes 0 and 2 each has a disliked count of 1
disliked_ct4 = [1, 0, 1]
start = time.time()
cp(data3, budget3, calorie_cap3, inventory3, disliked_ct4, allergies, chosen_meals=2)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


Solution status: OPTIMAL
Chosen recipe IDs: [0, 1]
  -> Chicken Pasta
  -> Spinach Salad
Extra purchased amounts:
Ingredient 1 (Pasta): 1.0
Ingredient 2 (Spinach): 1.0
Total cost used: 2.0 (Budget = 200)
Total Protein: 65
Total Calories: 700
Total Cholestrol: 100
Calories to Protein Ratio: 10.76923076923077
Calories to Cholesterol Ratio: 7.0

execution time: 0.0154 seconds


TEST 5: No Solution 

In the first version of this test, the calorie cap is too low for the recipe, meaning it won't be chosen

In the second version, the budget is insufficient to buy the needed ingredients, again leading to no recipes chosen

In [6]:
data5 = {
    "all_ingredients": [
        {"i_id": 0, "name": "Steak", "unit_price": 500},
    ],
    "recipes": [
        {
            "name": "Steak Dinner",
            "ingredients": [{"i_id": 0, "name": "Steak", "proportion": 1.0}],
            "nutrients": {"calories": 1200, "protein": 80, "cholesterol": 150}
        }
    ]
}

budget5 = 10
calorie_cap5 = 800 
inventory5 = [
    {"i_id": 0, "amount": 1.0},
]
disliked_ct5 = [0]
allergies = []
start = time.time()
cp(data5, budget5, calorie_cap5, inventory5, disliked_ct5, allergies, chosen_meals=1)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


No feasible solution found.

execution time: 0.0000 seconds


In [7]:
data5 = {
    "all_ingredients": [
        {"i_id": 0, "name": "Steak", "unit_price": 1500},
    ],
    "recipes": [
        {
            "name": "Steak Dinner",
            "ingredients": [{"i_id": 0, "name": "Steak", "proportion": 1.0}],
            "nutrients": {"calories": 1200, "protein": 80, "cholesterol": 150}
        }
    ]
}

budget5 = 10
calorie_cap5 = 1400 
inventory5 = [
    {"i_id": 0, "amount": 0.0},
]
disliked_ct5 = [0]
allergies = []
start = time.time()
cp(data5, budget5, calorie_cap5, inventory5, disliked_ct5, allergies, chosen_meals=1)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


No feasible solution found.

execution time: 0.0020 seconds


TEST 6: Medium Input (choose 5 recipes from 10)
here we can test a few different things, like how changing the input parameters will affect the results
1. no disliked foods, high budget ($30)
2. still no disliked foods, but lower the budget ($10)
3. dislike beef and broccoli, high budget
4. allergic to beef, dislike broccoli, high budget

In [8]:
data_large = {
    "all_ingredients": [
        {"i_id": 0, "name": "chicken", "unit_price": 150},
        {"i_id": 1, "name": "rice", "unit_price": 80},
        {"i_id": 2, "name": "broccoli", "unit_price": 120},
        {"i_id": 3, "name": "beef", "unit_price": 200},
        {"i_id": 4, "name": "pasta", "unit_price": 90},
        {"i_id": 5, "name": "tomato", "unit_price": 100},
        {"i_id": 6, "name": "cheese", "unit_price": 180},
        {"i_id": 7, "name": "lettuce", "unit_price": 110},
        {"i_id": 8, "name": "onion", "unit_price": 70},
        {"i_id": 9, "name": "beans", "unit_price": 130},
        {"i_id": 10, "name": "potato", "unit_price": 60},
    ],
    "recipes": [
        {"name": "chicken stir fry", "ingredients": [{"i_id": 0, "name": "chicken", "proportion": 0.5}, {"i_id": 2, "name": "broccoli", "proportion": 0.3}], "nutrients": {"calories": 500, "protein": 40, "cholesterol": 50}},
        {"name": "beef stew", "ingredients": [{"i_id": 3, "name": "beef", "proportion": 0.6}, {"i_id": 10, "name": "potato", "proportion": 0.4}], "nutrients": {"calories": 700, "protein": 60, "cholesterol": 90}},
        {"name": "veggie pasta", "ingredients": [{"i_id": 4, "name": "pasta", "proportion": 0.5}, {"i_id": 5, "name": "tomato", "proportion": 0.3}, {"i_id": 2, "name": "broccoli", "proportion": 0.2}], "nutrients": {"calories": 600, "protein": 20, "cholesterol": 10}},
        {"name": "chili beans", "ingredients": [{"i_id": 9, "name": "beans", "proportion": 0.7}, {"i_id": 8, "name": "onion", "proportion": 0.2}], "nutrients": {"calories": 500, "protein": 25, "cholesterol": 5}},
        {"name": "chicken salad", "ingredients": [{"i_id": 0, "name": "chicken", "proportion": 0.4}, {"i_id": 7, "name": "lettuce", "proportion": 0.3}], "nutrients": {"calories": 350, "protein": 30, "cholesterol": 40}},
        {"name": "cheesy pasta", "ingredients": [{"i_id": 4, "name": "pasta", "proportion": 0.5}, {"i_id": 6, "name": "cheese", "proportion": 0.5}], "nutrients": {"calories": 650, "protein": 25, "cholesterol": 60}},
        {"name": "baked potato", "ingredients": [{"i_id": 10, "name": "potato", "proportion": 0.9}], "nutrients": {"calories": 400, "protein": 8, "cholesterol": 0}},
        {"name": "beef tacos", "ingredients": [{"i_id": 3, "name": "beef", "proportion": 0.5}, {"i_id": 8, "name": "onion", "proportion": 0.3}], "nutrients": {"calories": 600, "protein": 45, "cholesterol": 70}},
        {"name": "rice and beans", "ingredients": [{"i_id": 1, "name": "rice", "proportion": 0.4}, {"i_id": 9, "name": "beans", "proportion": 0.4}], "nutrients": {"calories": 550, "protein": 18, "cholesterol": 5}},
        {"name": "pasta primavera", "ingredients": [{"i_id": 4, "name": "pasta", "proportion": 0.6}, {"i_id": 5, "name": "tomato", "proportion": 0.3}, {"i_id": 7, "name": "lettuce", "proportion": 0.1}], "nutrients": {"calories": 620, "protein": 22, "cholesterol": 10}},
    ]
}
inventory_large = [
    {"i_id": 0, "amount": 0.0},
    {"i_id": 1, "amount": 0.0},
    {"i_id": 2, "amount": 0.0},
    {"i_id": 3, "amount": 0.0},
    {"i_id": 4, "amount": 0.0},
    {"i_id": 5, "amount": 0.0},
    {"i_id": 6, "amount": 0.0},
    {"i_id": 7, "amount": 0.0},
    {"i_id": 8, "amount": 0.0},
    {"i_id": 9, "amount": 0.0},
    {"i_id": 10, "amount": 0.0}
]

disliked_ct_large = [0,0,0,0,0,0,0,0,0,0]

allergies = []
start = time.time()
cp(data_large, budget=30, calorie_cap=800, inventory=inventory_large, disliked_ct=disliked_ct_large, allergies=allergies, chosen_meals=5)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")


Solution status: OPTIMAL
Chosen recipe IDs: [0, 1, 3, 4, 7]
  -> chicken stir fry
  -> beef stew
  -> chili beans
  -> chicken salad
  -> beef tacos
Extra purchased amounts:
Ingredient 0 (chicken): 1.0
Ingredient 2 (broccoli): 1.0
Ingredient 3 (beef): 2.0
Ingredient 4 (pasta): 2.0
Ingredient 5 (tomato): 1.0
Ingredient 7 (lettuce): 1.0
Ingredient 8 (onion): 1.0
Ingredient 9 (beans): 2.0
Ingredient 10 (potato): 2.0
Total cost used: 15.1 (Budget = 30)
Total Protein: 200
Total Calories: 2650
Total Cholestrol: 255
Calories to Protein Ratio: 13.25
Calories to Cholesterol Ratio: 10.392156862745098

execution time: 0.0172 seconds


Now, if we lower the budget to $10, recipe 4 (chicken salad) gets replaced with recipe 6 (baked potato) since baked potato is cheaper ($0.54 cents versus $0.93) but the final result has a worse calorie to protein ratio since the amount of protein gets reduced also.

In [9]:
start = time.time()
cp(data_large, budget=10, calorie_cap=800, inventory=inventory_large, disliked_ct=disliked_ct_large, allergies=allergies, chosen_meals=5)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")

Solution status: OPTIMAL
Chosen recipe IDs: [0, 1, 3, 6, 7]
  -> chicken stir fry
  -> beef stew
  -> chili beans
  -> baked potato
  -> beef tacos
Extra purchased amounts:
Ingredient 0 (chicken): 1.0
Ingredient 2 (broccoli): 1.0
Ingredient 3 (beef): 2.0
Ingredient 8 (onion): 1.0
Ingredient 9 (beans): 1.0
Ingredient 10 (potato): 2.0
Total cost used: 9.9 (Budget = 10)
Total Protein: 178
Total Calories: 2700
Total Cholestrol: 215
Calories to Protein Ratio: 15.168539325842696
Calories to Cholesterol Ratio: 12.55813953488372

execution time: 0.0210 seconds


Now, what if we really dislike some of the ingredients? Say broccoli and beef :(

We see that (compared to baseline recipes), beef stew stays even though we dislike beef because it has so much protein, but beef tacos and chicken stir-fry are gone and replaced with rice and beans and pasta primavera

Additionally, note that the total cost increases from $15.10 to $15.90 and the total protein drops from 200g to 155g

In [10]:
disliked_ct_large = [1,1,1,0,0,0,0,1,0,0]

start = time.time()
cp(data_large, budget=30, calorie_cap=800, inventory=inventory_large, disliked_ct=disliked_ct_large, allergies=allergies, chosen_meals=5)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")

Solution status: OPTIMAL
Chosen recipe IDs: [1, 3, 4, 8, 9]
  -> beef stew
  -> chili beans
  -> chicken salad
  -> rice and beans
  -> pasta primavera
Extra purchased amounts:
Ingredient 0 (chicken): 1.0
Ingredient 1 (rice): 1.0
Ingredient 2 (broccoli): 1.0
Ingredient 3 (beef): 2.0
Ingredient 4 (pasta): 2.0
Ingredient 5 (tomato): 1.0
Ingredient 7 (lettuce): 1.0
Ingredient 8 (onion): 1.0
Ingredient 9 (beans): 2.0
Ingredient 10 (potato): 2.0
Total cost used: 15.9 (Budget = 30)
Total Protein: 155
Total Calories: 2720
Total Cholestrol: 150
Calories to Protein Ratio: 17.548387096774192
Calories to Cholesterol Ratio: 18.133333333333333

execution time: 0.0059 seconds


OK, but if we really hate beef we can tell a small lie and say we're actually allergic (yikes don't do this at a restaurant)
Now, notice that beef stew and beef tacos are no longer in the chosen recipes, but even though we still say we don't like broccoli which is in chicken stir fry, it is chosen since it has a lot of calories that are now "worth it" despite the user's dislike.

Here, the price actually goes down, probably since beef is quite expensive, but notice that the total protein once again drops to 135g (from baseline 200g and 155g with disliked but not allergic to beef)

In [11]:
allergies = ['beef']
disliked_ct_large = [1,0,1,0,0,0,0,0,0,0]
start = time.time()
cp(data_large, 30, 800, inventory_large, disliked_ct_large, allergies, 5)
end = time.time()
execution_time = end - start
print(f"\nexecution time: {execution_time:.4f} seconds")

Solution status: OPTIMAL
Chosen recipe IDs: [0, 3, 4, 8, 9]
  -> chicken stir fry
  -> chili beans
  -> chicken salad
  -> rice and beans
  -> pasta primavera
Extra purchased amounts:
Ingredient 0 (chicken): 1.0
Ingredient 1 (rice): 1.0
Ingredient 2 (broccoli): 1.0
Ingredient 4 (pasta): 2.0
Ingredient 5 (tomato): 1.0
Ingredient 7 (lettuce): 1.0
Ingredient 8 (onion): 1.0
Ingredient 9 (beans): 2.0
Total cost used: 10.7 (Budget = 30)
Total Protein: 135
Total Calories: 2520
Total Cholestrol: 110
Calories to Protein Ratio: 18.666666666666668
Calories to Cholesterol Ratio: 22.90909090909091

execution time: 0.0099 seconds


Finally, let's see what happens with our largest possible dataset of 100 recipes (sorry, we couldn't afford more recipes :()

Feel free to play around with the inputs in their respective files (in the /data folder), the only file that should not be changed is large_recipe_dataset since it comes from Spoonacular and must be in that specific format for the CP solver to use.

For reference, this is what we have initialized them to:
INVENTORY:
broccoli: 2
chicken breast: 1
pasta: 2
onion: 1
tomato: 1
potato: 2
eggs: 0.5

ALLERGIES:
peanuts
peanut butter
milk

DISLIKED:
cilantro
basil
olives
lemon

CAP: (calories/meal, # meals to return, max budget)
700,5,150


In [14]:
main("data/large_recipe_dataset.json", 
         "data/large_inventory.txt",
        "data/large_disliked.txt",
        "data/large_cap.txt",
        "data/large_allergies.txt")

Solution status: OPTIMAL
Chosen recipe IDs: [26, 33, 48, 69, 94]
  -> Best Minestrone Soup Recipe
  -> Easy Vegetable Beef Soup
  -> How to Make a Chicken Taco Crock Pot
  -> Delicious Creamy Lentils and Chestnuts Soup
  -> Mushroom, roasted tomato and garlic pasta
Extra purchased amounts:
Ingredient 1 (canned tomatoes): 3.0
Ingredient 3 (onion): 1.0
Ingredient 5 (parsley): 2.0
Ingredient 6 (garlic): 4.0
Ingredient 7 (ground beef i like): 2.0
Ingredient 10 (parmesan reggiano): 2.0
Ingredient 14 (thyme leaves): 3.0
Ingredient 17 (butter lettuce): 1.0
Ingredient 19 (capers): 1.0
Ingredient 20 (chicken stock): 1.0
Ingredient 22 (kidney beans): 2.0
Ingredient 23 (pepper flakes): 1.0
Ingredient 33 (kernal corn): 2.0
Ingredient 35 (spaghetti): 1.0
Ingredient 44 (rosemary): 1.0
Ingredient 45 (chilis): 1.0
Ingredient 71 (ziti pasta): 1.0
Ingredient 72 (celery seed): 1.0
Ingredient 75 (baby bella mushrooms): 1.0
Ingredient 77 (elbow macaroni): 1.0
Ingredient 80 (worcestershire sauce): 1.0
Ingre

Here is just an instance of a generic solver whose recipes database will be populated by your input, make sure to run python /preprocessing/pipeline.py to input the ingredients you want to base the recipes on.

It will take ~10 minutes for the recipe dataset to generate (due to API rate limits), so if you want to get fewer recipes for it to be faster, change line 250 in preprocessing/get_recipes.py to be     
recipes = fetch_enriched_recipes(my_ingredients, max_results={#results you want})


In [13]:
main("preprocessing/combined_recipe_data.json", 
         "preprocessing/inventory.txt",
        "preprocessing/disliked.txt",
        "preprocessing/cap.txt",
        "preprocessing/allergies.txt")

Solution status: OPTIMAL
Chosen recipe IDs: [1, 2, 3, 5, 10, 11, 14]
  -> Vegetarian Christmas wreath
  -> Homemade Broccoli Cheddar Soup
  -> Chicken Enchilada Salad Wraps
  -> Broccoli Cheddar Soup
  -> Buffalo Chicken Wings Wonton Wraps
  -> Broccoli Cheese Soup
  -> Bbq Chicken and Goat Cheese Ravioli
Extra purchased amounts:
Ingredient 1 (green onion): 6.0
Ingredient 2 (parmesan): 1.0
Ingredient 4 (butter): 1.0
Ingredient 5 (egg): 1.0
Ingredient 6 (powdered garlic): 3.0
Ingredient 7 (salmon): 2.0
Ingredient 8 (cheese): 2.0
Ingredient 9 (cherry tomatoes): 2.0
Ingredient 10 (bell pepper): 3.0
Ingredient 12 (dill): 1.0
Ingredient 13 (olives): 1.0
Ingredient 14 (juice of lemon): 1.0
Ingredient 15 (vegetable broth): 1.0
Ingredient 16 (yogurt): 1.0
Ingredient 17 (yukon gold potatoes): 2.0
Ingredient 18 (bay leaf): 1.0
Ingredient 19 (swanson premium chicken): 2.0
Ingredient 20 (cream): 1.0
Ingredient 21 (chili powder): 1.0
Ingredient 22 (cumin): 1.0
Ingredient 23 (cilantro): 1.0
Ingredie